### DISCUSSION TITLE
GeoTOP and uncertainty quantification

### DISCUSSION THREAD

**Discussion by JoostGevaert** [gempy-project/gempy#902] [Q&A] [answered]
Hi @Leguark,@flohorovicic,

I found the [gempy-geotop](https://github.com/gempy-project/gempy-geotop) (forked from https://github.com/terranigma-solutions/gempy-geotop-pilot) repo.
Great that this research is being done in the open!
Could you tell me something more about the project? Is it purely an R&D project to compare GemPy results with the GeoTOP model, or are you working together with the Dutch Geological Survey to see if GemPy can be used instead of [ISATIS](https://www.geovariances.com/en/software/isatis-neo-geostatistics-software/) for future GeoTOP iterations?

Besides, I read that both GeoTOP and GemPy use the notion of information entropy as a measure for the uncertainty in the geological model, which must be determined by running multiple simulations.
I was curious whether it's also possible to get a "cheap" quantification of uncertainty out of the variance of the scalar field. I'm not sure how to access the variance of the scalar field in GemPy, but my reasoning was that the scalar field is calculated through Kriging, and that therefore the variance of the scalar field should also be easy to calculate. The value the scalar field might not be a very useful number, but I thought that the variance of this scalar field might say something about the uncertainty of the geological model at a given point.

I'm looking forward to hearing from you!

**Comment by flohorovicic** [comment]
Hi @JoostGevaert,

yes, indeed - some nice things going on 😄.

mill comment on the information theory: this is going back to some time ago where we looked for a good way to visualise uncertainties in 3-D. Paper here: 

https://scholar.google.com.au/scholar?hl=de&as_sdt=0%2C5&q=Wellmann+Regenauer+information&oq=#d=gs_qabs&t=1716921215302&u=%23p%3DLqCtPEfuFG0J

For the calculation of the measure, you need an estimate of probability. And you are right: in our case obtained from multiple realisations. Theoretically, you could obtain it from variance, I would assume - but note that this would be an estimate of entropy from a continuous variable - a totally different beast.

Concerning the variance itself: in GemPy, we have a dual formulation implemented, which does not give you a direct estimate of variance. Gabriel Courrioux did this a while ago for the implementation in Geo modeller, though. I think it is only a conference paper, but probably interesting for you, if you are looking for it. But this would mean entering the depths of the implementation in GemPy, probably quite some work required.

**Reply by JoostGevaert** [reply]
Yes, when investigating both GemPy and GeoTOP I indeed stumbled upon that 2012 paper about using "information entropy" to quantify and visualize geological model uncertainty in 3D. Was pleased to see that your research was put into practice by the Dutch government :)

Thanks for your insight regarding the uncertainty quantification. What do you mean with "a dual formulation"?
I understood from the 2019 paper by Miguel that GemPy uses Kriging to interpolate the scalar field, and I always thought that Kriging gives you:
1. the interpolated value,
2. the variance of that estimation, which is based on the variogram model.

**Reply by JoostGevaert** [reply]
What I do still find complicated to explain in layman's terms is "information entropy". How do you do that?

I actually don't fully understand it myself either. Below some observations (please correct me if I'm wrong :) that I find hard to comprehend as a coherent whole:

A high information entropy value means high uncertainty, i.e. "bad". This is opposite to the probabilities, where a high value means low uncertainty, i.e. "good".

The information entropy of the GeoTOP model ranges from 0 to 1 (see [Ch 5.5.16 & 17 berekenen modelonzekerheid van lithoklasse & geologische eenheid](https://www.dinoloket.nl/sites/default/files/docs/geotop/R10758%20Totstandkomingsrapport%20GeoTOP%20-%20met%20nieuwe%20methodiek%20lagenmodellering.pdf)). The Entropy values in Fig. 9 of [GemPy 1.0: open-source stochastic geological modeling and inversion (Varga, Schaaf, Wellmann 2019)](https://gmd.copernicus.org/articles/12/1/2019/gmd-12-1-2019.pdf) range from 0 to 1.25.  
So, what do these values mean?  
What is / should be a typical range for information entropy?  
Is one free to normalize the probabilities before calculating the information entropy?

Lastly, based on Fig. 9 I understand that the information entropy is highest, i.e. "worst", on the boundaries between lithology classes. Which makes sense, because that is where the different experiments from the Monte Carlo method produce different lithology class results. However, the boundaries are exactly the places where we actually know most. As in, (most of) the input data is defined on these interfaces.   So, the infromation entropy doesn't tell us much about a lack of information?  
The variance of the scalar field would tell me something about a lack of information, right?

**Reply by flohorovicic** [reply, marked as answer]
Hi @JoostGevaert ,

Concerning the interpretation of IE, please have a look at the publication - there is the (fairly typical 😄) coin-flip example in it (if I recall right), which should provide you with the intuition. The extension to multiple outcomes is straight-forward.

Good text books are:
- MacKay Information theory and learning algorithms
- Cover & Thomas Information theory

Scaling is, in my opinion, not needed and also generally not a good idea. See also comments in our 2012 paper.

Concerning the last point: in geological models, the boundaries are exactly what we represent - so we can also only represent uncertainties related to the boundaries. If you have a model with additional components, e.g. geostatistical models inside the domain, you could, of course, use entropy to quantify uncertainties in these domains, as well.

**Reply by JoostGevaert** [reply]
Thanks a lot for your explanations, and patience! I should've read the 2012 paper before bothering again with all the questions. Will read it now :)
